glean-core/ios/Glean/Dispatchers.swift

```diff
@@ -0,0 +1,166 @@
+    private var queueInitialTasks = AtomicBoolean(true)
+
+    // This array will hold the queued initial tasks that are launched before Glean is initialized
+    lazy var preInitOperations: [Operation] = [Operation]()
+
+    // When true, jobs will be run synchronously
+    var testingMode = false
+
+    // Don't let other instances be created, we only want singleton access through the static `shared`
+    // property
+    private init() {}
+
+    /// This function launches a background `Operation`
+    ///
+    /// - parameters:
+    ///   * block: The block of code to execute, as a Closure that accepts no arugments and returns Void
+    ///
+    /// This function is used throughout Glean in order to launch tasks in the background, typically
+    /// recording of metrics, uploading of pings, etc.  This function returns an `Operation` which can
+    /// be used in conjunction with a completion handler to perform work after the task is complete.
+    ///
+    /// For example:
+    /// ```
+    /// Dispatchers.shared.launch {
+    ///     // Do some work here
+    /// }.completionBlock = {
+    ///     print("Done!")
+    /// }
+    /// ```
+    ///
+    /// You could also use the returned `Operation` to await that specific task like this:
+    /// ```
+    /// let operation = Dispatchers.shared.launch {
+    ///     // Do some work here
+    /// }
+    /// operation.waitUntilFinished()
+    func launch(block: @escaping () -> Void) -> Operation {
+        let operation = GleanOperation(block)
+
+        if queueInitialTasks.value {
+            // If we are queuing tasks, typically before Glean has been initialized, then we should
+            // just add the created Operation to the preInitOperations array, provided there are
+            // less than the max queued operations stored.
+            if preInitOperations.count < Constants.maxQueueSize {
+                preInitOperations.append(operation)
+
+                if testingMode {
+                    NSLog("\(Constants.logTag) : Task queued for execution in test mode")
+                } else {
+                    NSLog("\(Constants.logTag) : Task queued for execution and delayed until flushed")
+                }
+            } else {
+                NSLog("\(Constants.logTag) : Exceeded maximum queue size, discarding task")
+            }
+        } else {
+            // If we are not queuing initial tasks, we can go ahead and execute the Operation by
+            // adding it to the `operationQueue`
+            if testingMode {
+                // If we are in testing mode, go ahead and execute the operation and wait until it
+                // is finished before continuing to give us synchronous execution of the task.
+                operation.start()
+                operation.waitUntilFinished()
+            } else {
+                operationQueue.addOperation(operation)
+            }
+        }
+
+        return operation
```

Comment:
(Not critical to landing this, this is internal API so we control all execution paths)

To answer my own question from above to await something returned from `launch` you need to call `waitUntilFinished`.
But what if I call `start` on the thing that is returned? Is this returned operation a clone of what's in the queue and thus things get executed twice? Or is it the same object and the task in the queue will not be executed again later?

Reply:
You shouldn't be able to call `start()` on a `finished` operation, if I understand how the `Operation` class works. It needs to be in the `isReady` state.

Reply:
What happens then? An exception?

Reply:
I will look it up again to be sure, but I think it's a no-op.